notifications/pipeline_manager.py

```diff
@@ -61,10 +64,31 @@ def process(self, message):
         pipeline_step_pointer = self.pipeline_step_list.index(message.current_step)
         pipeline_step_pointer += 1
         message.current_step = self.pipeline_step_list[pipeline_step_pointer]  # Update the current step
-        message.history.append(PipelineHistory(message.current_step, 'Started'))
-        import_from_string(self.pipeline_step_list[pipeline_step_pointer]).process_message(message)
-        message.history.append(PipelineHistory(message.current_step, 'Completed'))
-        if message.current_step == self.pipeline_step_list[-1]:
-            logging.info("%s %s has been delivered.", message.name, message.uuid)
+        if pytz.utc.localize(datetime.utcnow()) > message.expiration_time:
+            logging.warning(
+                "%s %s has expired before step %s. It will be dropped.",
+                message.name, message.uuid, message.current_step
+                )
         else:
-            self.process(message)
+            logging.info(
+                "%s %s has not expired before step %s. It will be forwarded.",
+                message.name, message.uuid, message.current_step
+                )
+            message.history.append(PipelineHistory(message.current_step, 'Started'))
+            import_from_string(self.pipeline_step_list[pipeline_step_pointer]).process_message(message)
+            if pytz.utc.localize(datetime.utcnow()) > message.expiration_time:
+                logging.warning(
+                    "%s %s has expired after step %s. It will be dropped.",
+                    message.name, message.uuid, message.current_step
+                    )
+                message.history.append(PipelineHistory(message.current_step, 'Expired'))
+            else:
+                logging.info(
+                    "%s %s has not expired after step %s. It will be forwarded.",
+                    message.name, message.uuid, message.current_step
+                    )
```

Comment:
They differ slightly because they capture dropping before and after processing by the step but I am going to see if it can be improved.